# Ej 15

Obtenga la matriz de distancias euclídeas para todos los contenidos que están en Marte. ¿Cuáles son los dos contenidos que están a menor distancia?  

In [1]:
# importacion general de librerias y de visualizacion (matplotlib y seaborn)
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.neighbors import NearestNeighbors
from sklearn.metrics.pairwise import euclidean_distances 

%matplotlib inline

plt.style.use('default') # haciendo los graficos un poco mas bonitos en matplotlib
#plt.rcParams['figure.figsize'] = (20, 10)

sns.set(style="whitegrid") # seteando tipo de grid en seaborn

pd.options.display.float_format = '{:20,.2f}'.format # suprimimos la notacion cientifica en los outputs

import warnings
warnings.filterwarnings('ignore')

from datetime import datetime

In [2]:
geo_tags_db = pd.read_csv('./Wikipedia_dataset_2c2021/geo_tags.csv')

contenidos = pd.read_csv('./Wikipedia_dataset_2c2021/contents_text_sample.csv')

# 1. Busco los contenidos que estan en marte

In [3]:
contenido_marte = geo_tags_db[geo_tags_db['gt_globe']=='mars']
display(contenido_marte)

gt_id  gt_page_id gt_globe  gt_primary               gt_lat  \
1186     74461565     7473734     mars           0                16.90   
1187     74461566     7473734     mars           0                37.40   
1188     74461567     7473734     mars           0               -17.70   
1189     74461568     7473734     mars           0                16.30   
1190     74461569     7473734     mars           0               -28.70   
...           ...         ...      ...         ...                  ...   
486852  136540513      969836     mars           0                40.74   
515114  136586654     6623501     mars           1                38.27   
523403  136599374     8148400     mars           1                -2.12   
549564  136665032     9332614     mars           0                18.44   
551512  136667021     6105621     mars           1                -4.59   

                     gt_lon  gt_dim gt_type   gt_name gt_country gt_region  
1186                 141.70    1000     NaN   Corinto        NaN       NaN  
1187                 126.40    1000     NaN     Dilly        NaN       NaN  
1188                 199.90    1000     NaN  Gratteri        NaN       NaN  
1189                 125.90    1000     NaN    Tomini        NaN       NaN  
1190                 226.90    1000     NaN     Zumba        NaN       NaN  
...                     ...     ...     ...       ...        ...       ...  
486852               350.54    1000     NaN       NaN        NaN       NaN  
515114               224.98   10000    city       NaN        NaN       NaN  
523403               354.48    1000     NaN       NaN        NaN       NaN  
549564                77.45    1000     NaN       NaN        NaN       NaN  
551512               137.44    1000     NaN       NaN        NaN       NaN  

[1438 rows x 11 columns]

In [4]:
contenido_interes = contenidos.merge(contenido_marte, how='inner', left_on='id', right_on='gt_page_id' )
display(contenido_interes)

id                        title  \
0   1187405     Davies (cráter marciano)   
1   1389954              Centauri Montes   
2   1523483     Holden (cráter marciano)   
3   1538501            Cantoura (cráter)   
4   1538562             Calamar (cráter)   
5   1539222      Graff (cráter marciano)   
6   1539240              Hadley (cráter)   
7   1798092               Kufra (cráter)   
8   2951765                Azul (cráter)   
9   7709194     Hutton (cráter marciano)   
10  7720186    Janssen (cráter marciano)   
11  7818758      Marth (cráter marciano)   
12  8010820    Slipher (cráter marciano)   
13  9997383  Explorador de Marte Zhurong   

                                                 text      gt_id  gt_page_id  \
0   {{Ficha de cráter lunar |\n   nombre          ...  135795389     1187405   
1   {{Ficha de montaña\n| Nombre = Centauri Montes...  135809916     1389954   
2   {{Ficha de cráter espacial\n   |nombre   = Hol...  135818323     1523483   
3   {{Ficha de cráter espacial |\n   nombre   = Ca...  135819458     1538501   
4   {{Ficha de cráter espacial |\n   nombre   = Ca...  135819456     1538562   
5   {{Ficha de cráter espacial |\n   nombre   = Gr...  135819520     1539222   
6   {{Ficha de cráter espacial |\n   nombre   = Ha...  135819524     1539240   
7   \n{{Ficha de cráter lunar |\n   nombre        ...  135832417     1798092   
8   {{Ficha de cráter espacial\n| nombre = Azul\n|...  135871640     2951765   
9   {{Ficha de cráter espacial |\n   nombre   = Hu...  136142238     7709194   
10  {{Ficha de cráter espacial |\n   nombre   = Ja...  136142905     7720186   
11  {{Ficha de cráter espacial |\n   nombre   = Ma...  136146530     7818758   
12  {{Ficha de cráter espacial |\n   nombre   = Sl...  136154513     8010820   
13  {{Ficha de astronáutica\n|nombre              ...  136533856     9997383   

   gt_globe  gt_primary               gt_lat               gt_lon  gt_dim  \
0      mars           1                45.96                 0.09    1000   
1      mars           1               -38.67                95.52    1000   
2      mars           1               -26.04               325.98    1000   
3      mars           1                14.84               308.28    1000   
4      mars           1                18.27               305.13    1000   
5      mars           1               -21.18               153.81    1000   
6      mars           1               -19.26               156.97    1000   
7      mars           1                40.36               120.30    1000   
8      mars           1               -42.07               317.49    1000   
9      mars           1               -71.63               104.60    1000   
10     mars           1                 2.69                37.61    1000   
11     mars           1                12.94               356.55    1000   
12     mars           1               -47.34               275.54    1000   
13     mars           0                25.10               109.90    1000   

     gt_type gt_name gt_country gt_region  
0        NaN     NaN        NaN       NaN  
1   landmark     NaN        NaN       NaN  
2        NaN     NaN        NaN       NaN  
3        NaN     NaN        NaN       NaN  
4        NaN     NaN        NaN       NaN  
5        NaN     NaN        NaN       NaN  
6        NaN     NaN        NaN       NaN  
7        NaN     NaN        NaN       NaN  
8        NaN     NaN        NaN       NaN  
9        NaN     NaN        NaN       NaN  
10       NaN     NaN        NaN       NaN  
11       NaN     NaN        NaN       NaN  
12       NaN     NaN        NaN       NaN  
13       NaN     NaN        NaN       NaN

# 2. Limpio los textos

In [5]:
contenido_interes['texto_limpio'] = contenido_interes['text'].map(lambda x: str(x).replace('\n', '')
         .replace('.', '')
         .replace('|','')
         .replace('[','')
         .replace(']','')
         .replace('{','')
        .replace('}','')
)

In [6]:
corpus = contenido_interes['texto_limpio'].to_list()

vectorizer = CountVectorizer()

matriz_embedding = vectorizer.fit_transform(corpus)

In [7]:
matriz_euclediana = euclidean_distances(matriz_embedding)

pd.DataFrame(matriz_euclediana)

0                    1                    2   \
0                  0.00                39.64                67.54   
1                 39.64                 0.00                52.74   
2                 67.54                52.74                 0.00   
3                 16.82                38.86                64.99   
4                 16.97                39.51                65.95   
5                 23.62                37.22                58.52   
6                 26.74                38.11                58.38   
7                 16.73                39.36                65.58   
8                 18.33                38.22                66.76   
9                 21.59                35.68                58.23   
10                27.18                34.67                54.42   
11                20.98                36.21                59.32   
12                64.25                47.53                47.34   
13               268.49               244.68               227.50   

                     3                    4                    5   \
0                 16.82                16.97                23.62   
1                 38.86                39.51                37.22   
2                 64.99                65.95                58.52   
3                  0.00                10.34                18.08   
4                 10.34                 0.00                19.08   
5                 18.08                19.08                 0.00   
6                 21.86                22.74                15.33   
7                 10.91                11.40                18.65   
8                 17.80                18.17                23.96   
9                 22.47                23.07                23.24   
10                26.34                27.18                22.25   
11                20.86                21.35                19.18   
12                63.03                63.72                57.58   
13               265.90               266.74               260.30   

                     6                    7                    8   \
0                 26.74                16.73                18.33   
1                 38.11                39.36                38.22   
2                 58.38                65.58                66.76   
3                 21.86                10.91                17.80   
4                 22.74                11.40                18.17   
5                 15.33                18.65                23.96   
6                  0.00                22.52                27.18   
7                 22.52                 0.00                18.17   
8                 27.18                18.17                 0.00   
9                 25.71                22.89                24.17   
10                24.78                27.00                27.07   
11                22.74                21.21                22.98   
12                57.45                63.13                62.24   
13               259.20               266.58               265.61   

                     9                    10                   11  \
0                 21.59                27.18                20.98   
1                 35.68                34.67                36.21   
2                 58.23                54.42                59.32   
3                 22.47                26.34                20.86   
4                 23.07                27.18                21.35   
5                 23.24                22.25                19.18   
6                 25.71                24.78                22.74   
7                 22.89                27.00                21.21   
8                 24.17                27.07                22.98   
9                  0.00                23.56                20.54   
10                23.56                 0.00                20.27   
11                20.54                20.27                 0.00   
12              

# 3. Hallo la mínima distancia

In [8]:
min_distancia = matriz_euclediana[matriz_euclediana>0].min()
display(min_distancia)

10.344080432788601

# 4. Hallo los textos correspondientes mas cercanos 

In [9]:
np.where(matriz_euclediana==min_distancia)

(array([3, 4]), array([4, 3]))

In [10]:
contenido_interes.iloc[3].title

'Cantoura (cráter)'

In [11]:
contenido_interes.iloc[4].title

'Calamar (cráter)'

# Respuesta

Observamos que los contenidos que estan en marte y son mas cercanos (segun la distancia Euclidea) son Cantoura (cráter) y Calamar (cráter) y se encuentran a una distancia de 10.34.